src/simulator/halo_simulator.py

```diff
@@ -223,17 +228,94 @@ def simulated_venn_diagram_reach_by_spend(
         Args:
             spends:  The hypothetical spend vector, equal in length to
               the number of publishers.  spends[i] is the amount that is
-              spent with publisher i.
+              spent with publisher i. Note that publishers with 0 spends will
+              not be included in the Venn diagram reach.
             budget:  The amount of privacy budget that can be consumed while
               satisfying the request.
             max_frequency:  The maximum frequency for which to report reach.
         Returns:
-            A list of ReachPoint. Each reach point represents the mapping from 
+            A list of ReachPoint. Each reach point represents the mapping from
             the spends of a subset of publishers to the differentially private
             estimate of the number of people reached in this subset.
         """
         raise NotImplementedError()
 
+    def _form_venn_diagram_regions(
+        self, spends: List[float], max_frequency: int = 1
+    ) -> Tuple[Dict[int, Set], Dict[int, List]]:
+        """Form Venn diagram regions that contain k+ reaches
+
+        For each subset of publishers, computes k+ reaches for those users
+        who are reached by the publishers in that subset.
+
+        Args:
+            spends:  The hypothetical spend vector, equal in length to
+              the number of publishers.  spends[i] is the amount that is
+              spent with publisher i. Note that publishers with 0 spends will
+              not be included in the Venn diagram reach.
+            max_frequency:  The maximum frequency for which to report reach.
+        Returns:
+            pub_set_by_region:  Contains the mapping of a binary representation
+              to a set of publisher ids that are in the corresponding region.
+            regions:  Contains k+ reaches indexed by their binary representation.
+              The k+ reach for a given region is given as a list r[] where r[k]
+              is the number of people who were reached AT LEAST k+1 times.
+        """
+        # Get user counts by spend for each active publisher
+        user_counts_by_pub_id = {}
+        for pub_id, spend in enumerate(spends):
+            if not spend:
+                continue
+            user_counts_by_pub_id[pub_id] = self._publishers[
+                pub_id
+            ]._publisher_data.user_counts_by_spend(spend)
+
+        if len(user_counts_by_pub_id) > MAX_ACTIVE_PUBLISHERS:
```

Comment:
Is there a unit test that exercises this?

Reply:
No. I will add one. Thanks for pointing out.

Reply:
Done